compiler/src/parsing/parser.mly

```diff
@@ -411,9 +412,9 @@ id_vec:
   | lcaret lseparated_nonempty_list(comma, id_typ) comma? rcaret {$2}
 
 data_declaration:
-  | TYPE UIDENT id_vec? equal typ { DataDeclaration.abstract ~loc:(to_loc $loc) (mkstr $loc($2) $2) (Option.value ~default:[] $3) (Some $5) }
-  | ENUM UIDENT id_vec? data_constructors { DataDeclaration.variant ~loc:(to_loc $loc) (mkstr $loc($2) $2) (Option.value ~default:[] $3) $4 }
-  | RECORD UIDENT id_vec? data_record_body { DataDeclaration.record ~loc:(to_loc $loc) (mkstr $loc($2) $2) (Option.value ~default:[] $3) $4 }
+  | TYPE REC? UIDENT id_vec? equal typ { DataDeclaration.abstract ~loc:(to_loc $loc) (if Option.is_some $2 then Recursive else Nonrecursive) (mkstr $loc($3) $3) (Option.value ~default:[] $4) (Some $6) }
+  | ENUM REC? UIDENT id_vec? data_constructors { DataDeclaration.variant ~loc:(to_loc $loc) (if Option.is_some $2 then Recursive else Nonrecursive) (mkstr $loc($3) $3) (Option.value ~default:[] $4) $5 }
+  | RECORD REC? UIDENT id_vec? data_record_body { DataDeclaration.record ~loc:(to_loc $loc) (if Option.is_some $2 then Recursive else Nonrecursive) (mkstr $loc($3) $3) (Option.value ~default:[] $4) $5 }
```

Comment:
Does it make sense to make `REC?` it's own rule that does the option mapping?

Reply:
Also, I think we should default the argument in the helper and pass the option with `?` so maybe the rule isn't needed if you do those things

Reply:
Not completely sure what you mean with the second comment; isn't `?` syntax for optional arguments e.g. a parameter `~rec_flag=?` rather than default arguments e.g. `~rec_flag=Nonrecursive`?

Reply:
@alex-snezhko I pushed what I meant, please take a look.